homeassistant/components/scheduler/schedule.py

```diff
@@ -0,0 +1,334 @@
+        last_end_dt = last_start_dt + self.instance_duration
+        return last_start_dt <= now <= last_end_dt
+
+    @property
+    def expired(self) -> bool:
+        """Return whether the schedule has expired."""
+        now = dt_util.utcnow()
+
+        # If a recurrence exists and there's at least one instance of it in the future,
+        # we're not expired:
+        if self.recurrence and self.recurrence.after(now, inc=True):
+            return False
+
+        # If either the initial start datetime or the end datetime (if it exists) are in
+        # in the future, we're not expired:
+        if self.start_datetime >= now or (
+            self.end_datetime and self.end_datetime >= now
+        ):
+            return False
+
+        return True
+
+    @callback
+    def _get_next_instance_datetimes(
+        self,
+    ) -> Tuple[Optional[datetime], Optional[datetime]]:
+        """Get the next starting (and, optionally, ending) datetimes."""
+        if not self._initial_instance_scheduled:
+            self._initial_instance_scheduled = True
+            return (self.start_datetime, self.end_datetime)
+
+        if self.recurrence:
+            start = self.recurrence.after(dt_util.utcnow(), inc=True)
+
+            # The recurrence has reached its end:
+            if not start:
+                return (None, None)
+
+            if self.end_datetime:
+                end = start + self.instance_duration
+            else:
+                end = None
+
+            return (start, end)
+
+        return (None, None)
+
+    def as_dict(self) -> dict:
+        """Return the schedule as a dict."""
+        the_dict = {
+            CONF_ENTITY_ID: self.entity_id,
+            CONF_START_DATETIME: self.start_datetime.isoformat(),
+        }
+
+        if self.end_datetime:
+            the_dict[CONF_END_DATETIME] = self.end_datetime.isoformat()
+        else:
+            the_dict[CONF_END_DATETIME] = None
+
+        if self.recurrence:
+            the_dict[CONF_RECURRENCE] = str(self.recurrence)
+        else:
+            the_dict[CONF_RECURRENCE] = None
+
+        return the_dict
+
+    @callback
+    def async_cancel(self) -> None:
+        """Cancel the schedule."""
+        self.active_instance.async_cancel()
+        self.active_instance = None
+
+        if self._async_revert_listener:
+            self._async_revert_listener()
+            self._async_revert_listener = None
+        if self._async_trigger_listener:
+            self._async_trigger_listener()
+            self._async_trigger_listener = None
+
+    @callback
+    def async_schedule(self) -> None:
+        """Schedule the next instance."""
+        start_dt, end_dt = self._get_next_instance_datetimes()
+
+        if not start_dt:
+            _LOGGER.info("No more instances of schedule: %s", self)
+            return
+
+        self.active_instance = ScheduleInstance(self._hass, self.entity_id)
+
+        async def trigger(self, executed_at: datetime) -> None:
```

Comment:
Instead of defining them inline, make them class instance methods.

Reply:
This creates the conflict described [here](#discussion_r355141967).

If we kept these methods inline, we could accommodate the above _and_ ensure that we aren't storing (or trying to cancel) `self.active_instance` in the case where there's no end datetime.

Reply:
@balloob Let me know what you think of https://github.com/home-assistant/home-assistant/pull/28938/commits/740db983f515b607e9ee714c88735e589a8a9d72.

Reply:
@balloob I moved those to instances and created small, inline wrappers to accommodate the issue above. Let me know?